packages/hardhat/test/MyUSDEngine.ts

```diff
@@ -0,0 +1,606 @@
+      const initialBalance = await staking.getBalance(user1.address);
+
+      // Fast forward time by 1 year
+      await ethers.provider.send("evm_increaseTime", [365 * 24 * 60 * 60]);
+      await ethers.provider.send("evm_mine", []);
+
+      const finalBalance = await staking.getBalance(user1.address);
+      expect(finalBalance).to.equal(initialBalance);
+    });
+
+    it("Should accrue interest correctly over time", async function () {
+      const borrowRate = 1000; // 10% annual
+      const savingsRate = 800; // 8% annual
+      await rateController.setBorrowRate(borrowRate);
+      await rateController.setSavingsRate(savingsRate);
+
+      const initialBalance = await staking.getBalance(user1.address);
+
+      // Fast forward time by 1 year
+      await ethers.provider.send("evm_increaseTime", [365 * 24 * 60 * 60]);
+      await ethers.provider.send("evm_mine", []);
+
+      const finalBalance = await staking.getBalance(user1.address);
+      const expectedBalance = initialBalance + (initialBalance * BigInt(savingsRate)) / 10000n;
+
+      expect(finalBalance).to.be.closeTo(expectedBalance, ethers.parseEther("0.001"));
+    });
+
+    it("Should accrue interest proportionally over partial time periods", async function () {
+      const borrowRate = 1200; // 12% annual
+      const savingsRate = 900; // 9% annual
+      await rateController.setBorrowRate(borrowRate);
+      await rateController.setSavingsRate(savingsRate);
+
+      const initialBalance = await staking.getBalance(user1.address);
+
+      // Fast forward time by 6 months
+      await ethers.provider.send("evm_increaseTime", [182 * 24 * 60 * 60]); // ~6 months
+      await ethers.provider.send("evm_mine", []);
+
+      const finalBalance = await staking.getBalance(user1.address);
+      const expectedBalance = initialBalance + (initialBalance * BigInt(savingsRate) * 182n) / (365n * 10000n);
+
+      expect(finalBalance).to.be.closeTo(expectedBalance, ethers.parseEther("0.001"));
+    });
+
+    it("Should handle multiple interest accrual periods", async function () {
+      await rateController.setBorrowRate(600); // 6%
+      await rateController.setSavingsRate(400); // 4% annual
+
+      const initialBalance = await staking.getBalance(user1.address);
+
+      // Fast forward 3 months
+      await ethers.provider.send("evm_increaseTime", [91 * 24 * 60 * 60]);
+      await ethers.provider.send("evm_mine", []);
+
+      const midBalance = await staking.getBalance(user1.address);
+
+      // Change rate and fast forward another 3 months
+      await rateController.setBorrowRate(1000); // 10%
+      await rateController.setSavingsRate(700); // 7% annual
+      await ethers.provider.send("evm_increaseTime", [91 * 24 * 60 * 60]);
+      await ethers.provider.send("evm_mine", []);
+
+      const finalBalance = await staking.getBalance(user1.address);
+
+      expect(midBalance).to.be.gt(initialBalance);
+      expect(finalBalance).to.be.gt(midBalance);
```

Comment:
These tests are very detailed and great. I wish we could be more precise with this one, using closeTo instead of just gt.